NLC Integration to add search capability

kevan-ibm opened the pull request:
This PR adds a search capability to the hubot script.

kurtismgit committed:
Initial implementation of OpenWhisk NLC trainer action.

kurtismgit committed:
Initial implementation of OpenWhisk NLC trainer action.

kurtismgit committed:
Fix readme formatting

kevan-ibm committed:
Added search feature to look for files based on classifiers.

kurtismgit committed:
Train objectstorage NLC after BluePics provides alchemy tags.

kurtismgit committed:
fix lint issues with OpenWhisk action

kevan-ibm committed:
Added test coverage to the new search command

kurtismgit committed:
Add unit test for OpenWhisk nlcTrainer action

kevan-ibm committed:
Added a interval timer to clean up older NLC classifiers.

kevan-ibm committed:
Added more test cases for nlcHelper.

kevan-ibm committed:
Fit and finish
- Additional test coverage
- More comments

jlpadilla commented on OpenWhisk/actions/nlcTrainer.js:
```diff
@@ -0,0 +1,432 @@
+	const watson = require('watson-developer-cloud');
+	const Cloudant = require('cloudant');
+
+	var log4js = require('log4js');
```
We have been using `winston` for logging in other libraries where we don't have Hubot's logger.  I don't have a preference either way, but it would be good to standardize on using the same library across all our projects.

kurtismgit replied:
@jlpadilla I get the sense that winston is better.  I would have used winston here, but OpenWhisk is limited in terms of which modules your whisk actions can use.  Unfortunately winston isn't one of them.  So I had to use log4js.

Search "Node.js runtime environments" on this page to see list of available modules for whisk actions:  https://console.ng.bluemix.net/docs/openwhisk/openwhisk_reference.html

jlpadilla replied:
Ok, wasn't aware of this limitation.

kevan-ibm committed:
Updates to README and help for obcjectstorage search

jlpadilla commented:
`NLCHelper.js` belongs in `hubot-ibmcloud-cognitive-lib`.  The code is too similar to what's already in cognitive-lib.  I'd prefer to re-write cognitive-lib as a more general purpose library capable of handling the cases needed here.

kurtismgit commented on src/lib/env.js:
```diff
@@ -0,0 +1,68 @@
+
+settings.syncInterval = parseInt(settings.syncInterval, 10);
+settings.lowThreshold = parseFloat(settings.lowThreshold);
+settings.highThreshold = parseFloat(settings.highThreshold);
```
Not sure what syncInterval, lowThreshold and highThreshold are?  I don't see them mentioned anywhere else in the code.  Please remove if they aren't needed.

kevan-ibm replied:
I will remove those.

kurtismgit commented on src/lib/nlcHelper.js:
```diff
@@ -0,0 +1,346 @@
+ *        context.robot.logger = Hubot Logger
+ *        context.settings.nlc_url = Watson NLC API URL (OPTIONAL, defaults to https://gateway.watsonplatform.net/natural-language-classifier/api)
+ *        context.settings.nlc_username = Watson NLC username (REQUIRED)
+ *        context.settings.password = Watson NLC password (REQUIRED)
```
Typo in constructor comment.  It's nlc_password, not password.

kevan-ibm replied:
Fixed.

kurtismgit commented on src/lib/nlcHelper.js:
```diff
@@ -0,0 +1,346 @@
+ * Helper method to finds a classifier which is available (training completed)
+ * and with the most recent creation date.  If none are 'Available' then find
+ * the most recent classifier that started training.  If none are training,
+ * start the training.
```
comment is wrong, because it doesn't actually start the training

kevan-ibm replied:
Fixed.

kurtismgit commented on src/lib/nlcHelper.js:
```diff
@@ -0,0 +1,346 @@
+NLCHelper.prototype._getClassifier = function() {
+	let dfd = Promise.defer();
+
+	if (this.classifier_cache) {
```
We shouldn't be using classifier_cache.  At least not the way it's currently implemented.  The code running on the bot should be smart enough to detect new classifiers.  The cache will prevent this.

kevan-ibm replied:
Removed.  @jlpadilla The common util would also remove the cache.  The NLCManager would need to handle cache.

kurtismgit commented on src/scripts/object.search.js:
```diff
@@ -0,0 +1,210 @@
+			processObjectSearch(robot, res, parameters.searchphrase);
+		}
+		else {
+			robot.logger.error(`${TAG}: Error extracting container name from text [${res.message.text}].`);
```
This error print out should should say "Error extracting search phrase from text...", not container name.

kevan-ibm replied:
Fixed.

kevan-ibm committed:
Fixed id of command.

kevan-ibm committed:
Addressed concerns raised in PR
- Removed cache fields from class
- Removed unused env
- Fixed comments and error messages

kevan-ibm committed:
Fixed lint error.
